rewrite: rewrote the http client part with simplifications and easier… (#10)
ferrous-systems/espressif-trainings: book/src/03_3_2_http_client.md

$ cargo doc --open
```

✅ Add your [network credentials](02_4_hello_board.md) to the `cfg.toml` as previously.

✅ Open the prepared project skeleton in `intro/http-client/exercise`. We'll read through the code together.

## Making a connection

By default only unencrypted HTTP is available, which rather limits our options of hosts to connect to. We're going to use `http://neverssl.com/`.

In `esp-idf`, HTTP client connections are managed by `http::client::EspHttpClient` in the `esp-idf-svc` crate. It implements the `http::client::Client` trait from `embedded-svc`, which defines functions for [HTTP request methods](https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol#Request_methods) like `GET` or `POST`. This is a good time to have a look at the documentation you opened with `cargo doc --open`.

✅ Add the url `http://neverssl.com/` is in the main function. This is the address we will query.

> ✅ Create a new `EspHttpClient` with default values (Try to find the method `new_default()` in the documentation).

Reviewer: explanation: It's not necessary to check the doc if you already give the method away as there is no further configuration necessary. 

```suggestion
✅ Create a new `EspHttpClient` with default values. Look for a suitable instantiating method in the documentation.
```

Reviewer 2: "instantiating method" -> "constructor"

✅ Create a new `EspHttpClient` with default values.
✅ Create a new `EspHttpClient` with default values (Try to find the method `new_default()` in the documentation).

TODO currently no `docs.rs` documentation for the svc crates (ping Espressif about this, it's a known issue though)

Calling HTTP functions (e.g. `get(url)`) on this client returns an `EspHttpRequest`, which must be turned into a `Writer` to reflect the client's option to send some data alongside its request. This makes more sense with `POST` and `PUT` but must still be performed with `GET`. 
Calling HTTP functions (e.g. `get(url)`) on this client returns an `EspHttpRequest`, which must be turned into a `Writer` to reflect the client's option to send some data alongside its request. 

After this optional send step the `Writer` can be turned into a `Response` from which the received server output can be read:

> TODO explain as_ref(). Or find a better way to point people to it. 

Reviewer: Can you revert the changes around the as_ref() topic?
You can keep the your solution as an additional alternative way.

Author: I did not change the `solution/main.rs`, the solution without `as_ref()` is in `solution1.rs` for the moment.

The parameter passed to `into_writer` is the number of bytes the client intends to send. Here we are not trying to send anything. 

A successful response has [a status code in the 2xx range](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes).

✅ Verify the connection was successful.

✅ Return an Error if the status is not in the 2xx range.

```rust
match status {
        200..=299 => {
        }
        _ => anyhow::bail!("unexpected response code: {}", status),
    }
```

> The status error can be returned with [Anyhow](https://docs.rs/anyhow/latest/anyhow/index.html), who returns the result and manages internally any type `Error` that implements `std::error:Error`.

Reviewer 2: ```suggestion
The status error can be returned with [`anyhow`](https://docs.rs/anyhow/latest/anyhow/index.html), who returns the result and manages internally any type `Error` that implements `std::error:Error`.
```

Reviewer 2: on top of my formatting nit, I think the "who..." part of the sentence could use some rewording. Maybe something like "...the `anyhow`[...] crate which contains various functionality to simplify application-level error handling. It supplies a universal `anyhow::Result<T>`, wrapping the success (`Ok`) case in `T` and removing the need to specify the `Err` type, as long as every error you return implements `std::error::Error`.

> 

Reviewer: I think here it would be better not to give away the solution as `match` is pretty straightforward.
```suggestion
Use `match` to ensure the status is in the 2xx range.  Return an error with [Anyhow](https://docs.rs/anyhow/latest/anyhow/index.html) if the status is outside of this range.

✅ Turn your `response` into a `embedded_svc::io::Read` reader by calling `response.reader()` and read the received data chunk by chunk into a `u8` buffer using `reader.do_read(&mut buf)`. `do_read` returns the number of bytes read - you're done when this value is `0`.

✅ Report the total number of bytes read.

> ✅ Log the received data to the console. Hint, the response in the buffert is in bytes, so you might need [a method](https://doc.rust-lang.org/std/str/fn.from_utf8.html) to convert from bytes to `&str`.

Reviewer: explanation: "Hint" should not go into the instruction. If you think everyone needs this information, include it into the instruction. Otherwise a Hint section at the end of the page is better. 
"Might" sounds like you may or may not need it, this is a bit misleading as you really do need that. 
```suggestion
✅ Log the received data to the console. Use [from_utf8](https://doc.rust-lang.org/std/str/fn.from_utf8.html) to convert from bytes to `&str`.
```

Reviewer 2: I'd add a note that this assumes valid UTF-8 from the server which technically isn't true - the encoding is sometimes sent in the HTTP headers, often isn't. (And then you have MIME types on top of that...) UTF-8 is a good default assumption though.
